File: Week-02/Day-02/main.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}
type LinkedList struct {
	head *Node
}

func main() {
	list := LinkedList{}
	list.InsertAtEnd(71)
	list.InsertAtEnd(69)
	list.InsertAtEnd(22)
	list.InsertAtEnd(7)
	fmt.Println("Original list: ")
	list.PrintList()

	list.Reverse()
	fmt.Println("Reverse list: ")
	list.PrintList()

	// Creating a loop for testing
	list.InsertAtEnd(10)
	list.InsertAtEnd(20)
	list.InsertAtEnd(30)
	list.InsertAtEnd(40)
	//	list.head.next.next.next.next = list.head.next
	list.PrintList()

	if list.DetectLoop() {
		fmt.Println("Loop Detected")
	} else {
		fmt.Println("No loop Detected")
	}

	// Finding the Middle Element of a Linked List
	middle := list.FindMiddle()
	if middle != nil {
		fmt.Printf("Middle element: %d\n", middle.data)
	} else {
		fmt.Println("The list is empty")
	}
}

// Reverse: reverse the linked list
func (list *LinkedList) Reverse() {
	var prev, next *Node
	current := list.head
	for current != nil {
		next = current.next
		current.next = prev
		prev = current
		current = next
	}
	list.head = prev

	//
	middle := list.FindMiddle()
	if middle != nil {
		fmt.Printf("Middle Element: %d\n", middle.data)
	} else {
		fmt.Println("The list is empty")
	}
}

// InsertAtEnd: inserts a new node at the end of the list
func (list *LinkedList) InsertAtEnd(data int) {
	newNode := &Node{data: data}
	if list.head == nil {
		list.head = newNode
		return
	}

	last := list.head
	for last.next != nil {
		last = last.next
	}
	last.next = newNode
}

// PrintList  prints all the nodes in the list
func (list *LinkedList) PrintList() {
	temp := list.head
	for temp != nil {
		fmt.Printf("%d -> ", temp.data)
		temp = temp.next
	}
	fmt.Println("nil")
}

// DetectLoop detects if there is a loop in the linked list
func (list *LinkedList) DetectLoop() bool {
	slow, fast := list.head, list.head
	for slow != nil && fast != nil && fast.next != nil {
		slow = slow.next
		fast = fast.next.next
		if slow == fast {
			return true
		}
	}
	return false
}

// FindMiddle: Finds the middle element of the linked list
func (list *LinkedList) FindMiddle() *Node {
	slow, fast := list.head, list.head
	for fast != nil && fast.next != nil {
		slow = slow.next
		fast = fast.next.next
	}
	return slow
}

File: Week-02/Day-02/main.cpp
#include <iostream>
using namespace std;

class Node {
public:
  int data;
  Node* next;

  Node(int val) : data(val), next(nullptr) {}
};

class SinglyLinkedList {
private:
  Node* head;

public:
  SinglyLinkedList() : head(nullptr) {}

  void insertAtEnd(int val){ 
    Node* newNode = new Node(val);
    if(!head) {
      head = newNode;
      return;
    }
    Node* temp = head;
    while(temp->next){
      temp = temp->next;
    }
    temp->next = newNode;
  }

  // DeletefromBeginning
  void deleteFromBeginning(){
    if(!head) return;
    Node* temp = head;
    head = head->next;
    delete temp;
  }
 
// deletefromEnd:
  void deleteFromEnd(){
    if(!head) return;
    if(!head->next){
      delete head;
      head= nullptr;
      return;
    }
    Node* temp = head;
    while(temp->next && temp->next->next){
      temp = temp->next;
    }
    delete temp->next;
    temp->next = nullptr;
  }

  // deleteFromPosition
  void deleteFromPosition(int position){
    if (position == 0){
      deleteFromBeginning();
      return;
    }
    Node* temp = head;
      for(int i = 0; i < position-1 && temp; ++i){
      temp = temp->next;
    }
    if (!temp || !temp->next) return;
    Node* nodeToDelete = temp->next;
    temp->next = temp->next->next;
    delete nodeToDelete;
  }

  // PrintList:
  void printList(){
    Node *temp = head;
    while(temp) {
      cout << temp->data << " ";
      temp = temp->next;
    }
    cout << endl;
  }
};

int main (){
  SinglyLinkedList list;
  
  list.insertAtEnd(1);
  list.insertAtEnd(2);
  list.insertAtEnd(3);

  cout << " Initial List: ";
  list.printList();

  list.deleteFromBeginning();
  cout << "After deleting from beginning: ";
  list.printList();

  list.deleteFromEnd();
  cout << "After deleting from end: ";
  list.printList();

  list.deleteFromPosition(1);
  cout << "After deleting from position 1: ";
  list.printList();
  
  return 0;
}
